lib/ui/painting.dart

```diff
@@ -4263,6 +4272,111 @@ class FragmentProgram extends NativeFieldWrapperClass1 {
   external Handle _shader(_FragmentShader shader, Float32List floatUniforms, List<ImageShader> samplerUniforms);
 }
 
+/// A [Shader] generated from a [FragmentProgram].
+///
+/// Instances of this class can be obtained from the
+/// [FragmentProgram.fragmentShader] method. The float uniforms list is
+/// initialized to the size expected by the shader and is zero-filled. Uniforms
+/// of float type can then be set by calling [setFloat]. Sampler uniforms are
+/// set by calling [setSampler].
+///
+/// A [FragmentShader] can be re-used, and this is an efficient way to avoid
+/// allocating and re-initializing the uniform buffer and samplers. However,
+/// if two [FragmentShader] objects with different float uniforms or samplers
+/// are required to exist simultaneously, they must be obtained from two
+/// different calls to [FragmentProgram.fragmentShader].
+///
+/// Consider a fragment shader with uniforms like the following:
+///
+/// ```glsl
+/// layout (location = 0) uniform float a;
+/// layout (location = 1) uniform vec2 b;
+/// layout (location = 2) uniform vec3 c;
+/// layout (location = 3) uniform mat2x2 d;
+/// layout (location = 4) uniform sampler2d s;
+/// ```
+///
+/// If this shader is loaded into a [FragmentProgram] object `program`, a
+/// [FragmentShader] object can be constructed as follows:
+/// ```dart
+/// ImageShader sampler;
+/// final FragmentShader shader = program.fragmentShader()
+/// ..setFloat(0, 0.0)
+/// ..setFloat(1, 1.0)
+/// ..setFloat(2, 2.0)
+/// ..setFloat(3, 3.0)
+/// ..setFloat(4, 4.0)
+/// ..setFloat(5, 5.0)
+/// ..setFloat(6, 6.0)
+/// ..setFloat(7, 7.0)
+/// ..setFloat(8, 8.8)
+/// ..setFloat(9, 9.0)
+/// ..setSampler(0, sampler);
+///
+/// Paint paint = Paint()..shader = shader;
+/// ```
+///
+/// The uniforms will then be set as follows:
+///
+/// a: 0.0
+/// b: [1.0, 2.0]
+/// c: [3.0, 4.0, 5.0]
+/// d: [6.0, 7.0, 8.0, 9.0] // 2x2 matrix in column-major order
+///
+/// On subsequent frames, if only the uniform `a` is required to vary, then
+/// only `shader.setFloat(0, newValue)` needs to be called. The other uniforms
+/// will retain their values. The sampler uniforms will also persist in the same
+/// way.
+class FragmentShader extends Shader {
```

Comment:
Another approach, if the ergonomics of `setFloat` aren't to your liking. Just create a view of the original object:

```
class Float32Uniforms {
  FragmentShader? _instance;
  
  operator[]=(int index, double value) {
     assert(disposedStuff);
     _instance._setFloat(index, value);
   }
}

class FragmentShader {
  Float32Uniforms _floatUniforms;
  Float32Uniforms get floatUniforms => _floatUniforms;
  
  void dispose() {
    _floatUniforms._instance = null;
  }
}
```

```

Reply:
I trust the VM's ability to inline the simple wrapper that's currently in the PR. Adding more indirection would make me start to worry...

Reply:
no prob. might be possible to do "for free" someday with views when/if they come out:

```dart
view Float32Uniforms on FragmentShader {
  void operator[]=(int index, double value) => this.setFloat(index, value);
}

class FragmentShader {
  Float32Uniforms get floatUniforms => this as Float32Uniforms;
}

```